File: solutions/0005-longest-palindromic-substring/longest-palindromic-substring.cpp
// Given a string s, return the longest palindromic substring in s.
//
//  
// Example 1:
//
//
// Input: s = "babad"
// Output: "bab"
// Note: "aba" is also a valid answer.
//
//
// Example 2:
//
//
// Input: s = "cbbd"
// Output: "bb"
//
//
// Example 3:
//
//
// Input: s = "a"
// Output: "a"
//
//
// Example 4:
//
//
// Input: s = "ac"
// Output: "a"
//
//
//  
// Constraints:
//
//
// 	1 <= s.length <= 1000
// 	s consist of only digits and English letters (lower-case and/or upper-case),
//


class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        string tor = "";
        vector<vector<int>> dp(n, vector<int>(n, 0)); 
        for (int l = 1; l <= n; l++){
            for (int i = 0; i + l - 1 < n; i++){
                int j = i + l - 1;
                if (i == j) {
                    dp[i][j] = 1;
                }else{
                    dp[i][j] = (s[i] == s[j] && (l == 2 || dp[i + 1][j - 1]))?1:0;
                }
                if (dp[i][j] && l > tor.size()){
                    tor = s.substr(i, l);
                }
            }
        }
        return tor;
    }
};

File: solutions/0005-longest-palindromic-substring/longest-palindromic-substring.py
# Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.
#
# Example 1:
#
#
# Input: "babad"
# Output: "bab"
# Note: "aba" is also a valid answer.
#
#
# Example 2:
#
#
# Input: "cbbd"
# Output: "bb"
#
#


class Solution:
    def longestPalindrome(self, s: str) -> str:
        assert('$' not in s and '^' not in s and '#' not in s)
        if s == '':
            return ''
        t = '^#' + "#".join(s) + "#$"
        c = 0
        d = 0
        #les rayons du palindrom cnetre en j
        P = [0 for _ in range(len(t))]
        for  i in range(1, len(t) - 1):
            mirroir = 2 * c - i
            P[i] = max(0, min(d - i, P[mirroir]))
            
            while t[i + 1 + P[i]] == t[i - 1 - P[i]]:
                P[i] += 1

            if i + P[i] > d:
                c = i
                d = i + P[i]
        (k, i) = max((P[i], i) for i in range(1, len(t) - 1))
        (start, end) = ((i - k) // 2, (i + k) // 2)
        return s[start:end]
                    
